### PR TITLE
Use pointer to async_send directly instead of a wrapper function

### LINKED ISSUE #1314: Segfault with CUDA in a sysimage
Workaround: take CUDA.jl out of the sysimage 😄. Also seems to work on Julia 1.6.5 (but not 1.7.1).

Cause: during sysimage creation we create broken code, according to @vchuravy who helped me try to debug a segfault on #gpu on Slack. He pointed to

https://github.com/JuliaGPU/CUDA.jl/blob/349707790ccfbdbd83de0540ffdb75edcb179c21/lib/cudadrv/execution.jl#L143

and 
```julia
julia> @code_llvm CUDA.async_send(C_NULL)
;  @ /root/.julia/packages/CUDA/Rhl18/lib/cudadrv/execution.jl:143 within `async_send`
define i32 @julia_async_send_1861(i64 zeroext %0) #0 {
top:
  %1 = call i32 inttoptr (i64 139633100932974 to i32 (i64)*)(i64 %0)
  ret i32 %1
}

```

and said

> You are reading from TLS on a foreign thread which is illegal

I don't understand most of that to be honest.

This is with CUDA#master (or v3.6.1), Julia 1.7.1, PackageCompiler v1.7.7.

---

Debugger session:

```bash
[New Thread 0x7f8ae0a22700 (LWP 660)]
[Thread 0x7f8ae0a22700 (LWP 660) exited]
[New Thread 0x7f8ae0a22700 (LWP 661)]

Thread 20 "julia" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7f8ae0a22700 (LWP 661)]
0x00007f8bab618905 in jlcapi_async_send_69899 () from /deps.so
(gdb) bt
#0  0x00007f8bab618905 in jlcapi_async_send_69899 () from /deps.so
#1  0x00007f8b30e2045a in ?? ()
   from /root/.julia/artifacts/b618eadab05abb3bef21da981a51b6f4f0183b27/lib/libcuda.so
#2  0x00007f8b30e88106 in ?? ()
   from /root/.julia/artifacts/b618eadab05abb3bef21da981a51b6f4f0183b27/lib/libcuda.so
#3  0x00007f8bd5628609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#4  0x00007f8bd554f293 in clone () from /usr/lib/x86_64-linux-gnu/libc.so.6
```
```bash
(gdb) disassemble jlcapi_async_send_69899
Dump of assembler code for function jlcapi_async_send_69899:
   0x00007f8bab6188c0 <+0>:     push   %r15
   0x00007f8bab6188c2 <+2>:     push   %r14
   0x00007f8bab6188c4 <+4>:     push   %r13
   0x00007f8bab6188c6 <+6>:     push   %r12
   0x00007f8bab6188c8 <+8>:     push   %rbx
   0x00007f8bab6188c9 <+9>:     sub    $0x20,%rsp
   0x00007f8bab6188cd <+13>:    mov    %rdi,%r14
   0x00007f8bab6188d0 <+16>:    xorps  %xmm0,%xmm0
   0x00007f8bab6188d3 <+19>:    movaps %xmm0,(%rsp)
   0x00007f8bab6188d7 <+23>:    movq   $0x0,0x10(%rsp)
   0x00007f8bab6188e0 <+32>:    mov    0x1e902459(%rip),%rax        # 0x7f8bc9f1ad40 <jl_tls_offset.real>
   0x00007f8bab6188e7 <+39>:    test   %rax,%rax
   0x00007f8bab6188ea <+42>:    je     0x7f8bab618999 <jlcapi_async_send_69899+217>
   0x00007f8bab6188f0 <+48>:    mov    %fs:0x0,%rcx
   0x00007f8bab6188f9 <+57>:    mov    (%rcx,%rax,1),%rbx
   0x00007f8bab6188fd <+61>:    movq   $0x4,(%rsp)
=> 0x00007f8bab618905 <+69>:    mov    (%rbx),%rax
--Type <RET> for more, q to quit, c to continue without paging--
```

### PR DESCRIPTION
**Pull request by vchuravy**
Fixes #1314

@ericphanson can you check that this fixes your issue?

### CODE REVIEW

**Comment by ericphanson** [comment]
Sure, I'll give it a try

**Comment by ericphanson** [comment]
it crashes at the same place it used to segfault, saying
```julia
ERROR: LoadError: could not load symbol "async_send":
julia: undefined symbol: async_send
Stacktrace:
  [1] launch(f::CUDA.var"#13#14"{Base.Event}; stream::CuStream)
    @ CUDA ~/.julia/packages/CUDA/ZRjZ5/lib/cudadrv/execution.jl:157
  [2] nonblocking_synchronize
    @ ~/.julia/packages/CUDA/ZRjZ5/lib/cudadrv/stream.jl:157 [inlined]
  [3] macro expansion
    @ ~/.julia/packages/CUDA/ZRjZ5/src/array.jl:409 [inlined]
  [4] macro expansion
    @ ~/.julia/packages/CUDA/ZRjZ5/src/state.jl:182 [inlined]
  [5] unsafe_copyto!(dest::Matrix{Float32}, doffs::Int64, src::CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, soffs::Int64, n::Int64)
    @ CUDA ~/.julia/packages/CUDA/ZRjZ5/src/array.jl:405
  [6] copyto!
    @ ~/.julia/packages/CUDA/ZRjZ5/src/array.jl:359 [inlined]
  [7] copyto!
    @ ~/.julia/packages/CUDA/ZRjZ5/src/array.jl:363 [inlined]
  [8] copyto_axcheck!(dest::Matrix{Float32}, src::CuArray{Float32, 2, CUDA.Mem.DeviceBuffer})
    @ Base ./abstractarray.jl:1104
  [9] Array
    @ ./array.jl:563 [inlined]
 [10] Array
    @ ./boot.jl:481 [inlined]
 [11] convert
    @ ./array.jl:554 [inlined]
 [12] adapt_storage
    @ ~/.julia/packages/Adapt/wASZA/src/arrays.jl:6 [inlined]
 [13] adapt_structure
    @ ~/.julia/packages/Adapt/wASZA/src/Adapt.jl:42 [inlined]
 [14] adapt
    @ ~/.julia/packages/Adapt/wASZA/src/Adapt.jl:40 [inlined]
 [15] adapt_storage
    @ ~/.julia/packages/Flux/BPPNj/src/functor.jl:75 [inlined]
 [16] adapt_structure
    @ ~/.julia/packages/Adapt/wASZA/src/Adapt.jl:42 [inlined]
 [17] adapt
    @ ~/.julia/packages/Adapt/wASZA/src/Adapt.jl:40 [inlined]
 [18] #134
    @ ~/.julia/packages/Flux/BPPNj/src/functor.jl:115 [inlined]
 [19] fmap(f::Flux.var"#134#135", x::CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}; exclude::typeof(Functors.isleaf), walk::typeof(Functors._default_walk), cache::IdDict{Any, Any})
    @ Functors ~/.julia/packages/Functors/hIysk/src/functor.jl:121
 [20] fmap
    @ ~/.julia/packages/Functors/hIysk/src/functor.jl:120 [inlined]
 [21] cpu(x::CuArray{Float32, 2, CUDA.Mem.DeviceBuffer})
    @ Flux ~/.julia/packages/Flux/BPPNj/src/functor.jl:115
[truncated where it reaches my code]
```

File: lib/cudadrv/execution.jl [hunk 1]
```diff
@@ -149,7 +149,12 @@ function launch(f::Base.Callable; stream::CuStream=stream())
     end
     # FIXME: protect this from GC
 
-    callback = @cfunction(async_send, Cint, (Ptr{Cvoid},))
+    # callback = @cfunction(async_send, Cint, (Ptr{Cvoid},))
+    # See https://github.com/JuliaGPU/CUDA.jl/issues/1314.
+    # and https://github.com/JuliaLang/julia/issues/43748
+    # TL;DR We are not allowed to cache `async_send` in the sysimage
+    # so instead let's just pull out the function pointer and pass it instead.
+    callback = cglobal(:async_send)
```
Comment (vchuravy) [inline lib/cudadrv/execution.jl]:
```suggestion
    callback = cglobal(:uv_async_send)
```

**Comment by vchuravy** [comment]
> it crashes at the same place it used to segfault, saying 

I was a silly goose.

**Comment by ericphanson** [comment]
Success! No segfault or other issues

**Comment by maleadt** [comment]
Great, thanks!

**Comment by ericphanson** [comment]
could we get a release with this fix?

**Comment by maleadt** [comment]
Yes, I'll work on a release today.

**Comment by ericphanson** [comment]
thanks!

**Pull request merged**